File: cpp/src/refine.cpp
#include "refine.hpp"

#include <algorithm>
#include <cmath>

namespace autolocalize {

std::vector<double> frange(double start, double stop, double step) {
  std::vector<double> values;
  if (step <= 0.0) {
    values.push_back(start);
    return values;
  }
  for (double v = start; v <= stop + step * 0.5; v += step) {
    values.push_back(v);
  }
  return values;
}

RefineResult search_grid(const PoseScorerCore& scorer, double x, double y,
                         double theta, double translation_step,
                         double translation_span, double rotation_step,
                         double rotation_span) {
  RefineResult best{x, y, theta, scorer.score_endpoints(x, y, theta)};

  const auto dx_vals = frange(-translation_span, translation_span, translation_step);
  const auto dy_vals = frange(-translation_span, translation_span, translation_step);
  const auto dtheta_vals =
      frange(-rotation_span, rotation_span, rotation_step);

  for (double dx : dx_vals) {
    for (double dy : dy_vals) {
      for (double dtheta : dtheta_vals) {
        if (dx == 0.0 && dy == 0.0 && dtheta == 0.0) {
          continue;
        }
        const double cx = x + dx;
        const double cy = y + dy;
        const double ctheta = normalize_angle(theta + dtheta);
        const double score = scorer.score_endpoints(cx, cy, ctheta);
        if (score > best.score) {
          best = {cx, cy, ctheta, score};
        }
      }
    }
  }

  return best;
}

RefineResult refine_multiscale(const PoseScorerCore& scorer, double x, double y,
                               double theta, double translation_span,
                               double rotation_span) {
  RefineResult best{x, y, theta, scorer.score_endpoints(x, y, theta)};

  struct Scale {
    double t_span;
    double t_step;
    double r_span;
    double r_step;
  };

  const Scale scales[] = {
      {std::max(translation_span, 0.35), 0.12, std::max(rotation_span, 0.45),
       0.12},
      {0.15, 0.06, 0.25, 0.06},
      {0.06, 0.03, 0.12, 0.04},
  };

  for (const Scale& scale : scales) {
    const RefineResult candidate = search_grid(
        scorer, best.x, best.y, best.theta, scale.t_step, scale.t_span,
        scale.r_step, scale.r_span);
    if (candidate.score > best.score) {
      best = candidate;
    }
  }

  return best;
}

}  // namespace autolocalize

File: autolocalize/localization/refine.py
from __future__ import annotations

from autolocalize.geometry.pose import Pose2D
from autolocalize.localization.fast_grid import PoseScorer


def refine_pose(
    scorer: PoseScorer,
    pose: Pose2D,
    *,
    translation_span: float = 0.2,
    rotation_span: float = 0.35,
) -> tuple[Pose2D, float]:
    """Local grid search around a pose to maximize endpoint score."""
    return _search_grid(
        scorer,
        pose,
        translation_step=0.08,
        translation_span=translation_span,
        rotation_step=0.08,
        rotation_span=rotation_span,
    )


def refine_pose_quick(
    scorer: PoseScorer,
    pose: Pose2D,
    *,
    translation_span: float = 0.16,
    rotation_span: float = 0.28,
) -> tuple[Pose2D, float]:
    """Coarser, smaller local search for adaptive tier 1/2 (fewer evaluations)."""
    return _search_grid(
        scorer,
        pose,
        translation_step=0.1,
        translation_span=translation_span,
        rotation_step=0.1,
        rotation_span=rotation_span,
    )


def refine_pose_multiscale(
    scorer: PoseScorer,
    pose: Pose2D,
    *,
    translation_span: float = 0.2,
    rotation_span: float = 0.35,
) -> tuple[Pose2D, float]:
    """Coarse-to-fine grid search for larger initial pose error."""
    best_pose, best_score = pose, scorer.score_fast(pose)
    scales = (
        (max(translation_span, 0.35), 0.12, max(rotation_span, 0.45), 0.12),
        (0.15, 0.06, 0.25, 0.06),
        (0.06, 0.03, 0.12, 0.04),
    )
    for t_span, t_step, r_span, r_step in scales:
        candidate, score = _search_grid(
            scorer,
            best_pose,
            translation_step=t_step,
            translation_span=t_span,
            rotation_step=r_step,
            rotation_span=r_span,
        )
        if score > best_score:
            best_pose, best_score = candidate, score
    return best_pose, best_score


def _search_grid(
    scorer: PoseScorer,
    pose: Pose2D,
    *,
    translation_step: float,
    translation_span: float,
    rotation_step: float,
    rotation_span: float,
) -> tuple[Pose2D, float]:
    best_pose = pose
    best_score = scorer.score_fast(pose)

    for dx in _frange(-translation_span, translation_span, translation_step):
        for dy in _frange(-translation_span, translation_span, translation_step):
            for dtheta in _frange(-rotation_span, rotation_span, rotation_step):
                if dx == 0.0 and dy == 0.0 and dtheta == 0.0:
                    continue
                candidate = Pose2D(
                    x=pose.x + dx,
                    y=pose.y + dy,
                    theta=Pose2D.normalize_angle(pose.theta + dtheta),
                )
                score = scorer.score_fast(candidate)
                if score > best_score:
                    best_score = score
                    best_pose = candidate

    return best_pose, best_score


def _frange(start: float, stop: float, step: float) -> list[float]:
    if step <= 0:
        return [start]
    values: list[float] = []
    v = start
    while v <= stop + step * 0.5:
        values.append(v)
        v += step
    return values
